### PR TITLE
ILabelStatement Refactor

### PR DESCRIPTION
**Pull request by 333fred**
Fixes https://github.com/dotnet/roslyn/issues/21285. Tagging @dotnet/analyzer-ioperation @dotnet/roslyn-compiler @AlekseyTs for review.

### CODE REVIEW

**Review by mavasani — APPROVED** [review]

File: src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb [hunk 1]
```diff
@@ -1061,19 +1061,19 @@ Namespace Microsoft.CodeAnalysis.Semantics
...
             Dim constantValue As [Optional](Of Object) = New [Optional](Of Object)()
             Dim isImplicit As Boolean = boundLabelStatement.WasCompilerGenerated
-            Return New LazyLabelStatement(label, labeledStatement, _semanticModel, syntax, type, constantValue, isImplicit)
+            Return New LazyLabeledStatement(label, statement, _semanticModel, syntax, type, constantValue, isImplicit)
         End Function
 
         Private Function CreateBoundGotoStatementOperation(boundGotoStatement As BoundGotoStatement) As IBranchStatement
             Dim target As ILabelSymbol = boundGotoStatement.Label
-            Dim branchKind As BranchKind = BranchKind.GoTo
+            Dim branchKind As BranchKind = branchKind.GoTo
```
Comment (cston) [inline src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb]: `BranchKind` here and below.

**Review by cston — APPROVED** [review]

File: src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb (hunk 1 above)
Comment (333fred) [inline src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb]: I'm seriously not changing these explicitly; why is the autoformatter renaming these?

File: src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb [hunk 2]
```diff
@@ -286,7 +286,7 @@ Namespace Microsoft.CodeAnalysis.Semantics
         End Function
 
         Private Function CreateBoundMeReferenceOperation(boundMeReference As BoundMeReference) As IInstanceReferenceExpression
-            Dim instanceReferenceKind As InstanceReferenceKind = If(boundMeReference.WasCompilerGenerated, InstanceReferenceKind.Implicit, InstanceReferenceKind.Explicit)
+            Dim instanceReferenceKind As InstanceReferenceKind = If(boundMeReference.WasCompilerGenerated, instanceReferenceKind.Implicit, instanceReferenceKind.Explicit)
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb]: >instanceReferenceKind [](start = 107, length = 21)

This looks like an unintentional change in casing.

Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb]: >instanceReferenceKind [](start = 139, length = 21)

Here too.

**Comment by AlekseyTs** [comment]
Either I missed them, or we don't have a single test with non-null underlying statement.

**Comment by AlekseyTs** [comment]
Done with review pass.

**Comment by 333fred** [comment]
@AlekseyTs VB doesn't have any, but C# has a couple [here](https://github.com/dotnet/roslyn/blob/features/ioperation/src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IForLoopStatement.cs#L914) and [here](https://github.com/dotnet/roslyn/blob/features/ioperation/src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IWhileUntilLoopStatement.cs#L437). Do you want me to add one or two for VB?

**Comment by AlekseyTs** [comment]
> VB doesn't have any, but C# has a couple here and here. Do you want me to add one or two for VB?

I think it would be great to have at least one test like that for VB.

**Comment by 333fred** [comment]
@AlekseyTs looking at the BoundLabelStatement, it appears that VB doesn't have the concept of a label having an underlying statement. I'll add a simple test with a label other than the implicit exit, but Statement will always be null in VB. I'll also update the comment to note that.

**Comment by 333fred** [comment]
Windows debug failure is https://github.com/dotnet/roslyn/issues/21588.

**Comment by 333fred** [comment]
Integration test failure is https://github.com/dotnet/roslyn/issues/21421.

**Comment by 333fred** [comment]
@dotnet-bot retest windows_debug_unit64_prtest please

**Comment by 333fred** [comment]
@dotnet-bot retest windows_release_vs-integration_prtest please

**Comment by 333fred** [comment]
Ping for reviews.

**Review by AlekseyTs — APPROVED** [review]
LGTM

**Comment by 333fred** [comment]
Ping @dotnet/analyzer-ioperation for another review.

**Comment by 333fred** [comment]
Ping @dotnet/analyzer-ioperation again, I need another review please.

**Review by CyrusNajmabadi — APPROVED** [review]

**Comment by 333fred** [comment]
Integration test failure is https://github.com/dotnet/roslyn/issues/21125.

**Comment by 333fred** [comment]
@dotnet-bot retest windows_debug_vs-integration_prtest please.
